README.md
# Prayer Time Dart Package

[![Pub](https://img.shields.io/pub/v/salat.svg)](https://pub.dev/packages/salat)

This Dart package provides functionality for calculating accurate Islamic prayer times. It is a translation of the original [Salat](https://github.com/zainhussaini/salat/) repository, bringing the power and flexibility of prayer time calculations to the Dart ecosystem.

## Features

- Calculation of precise prayer times based on different calculation methods.
- Support for a wide range of time zones and locations.
- Calculation of additional prayer-related timings such as sunrise and sunset.
- Customizable calculation settings to accommodate various calculation conventions and preferences.
- Easy-to-use API for retrieving prayer times for a specific date, location, and time zone.
- Flexible integration options with other Dart projects or frameworks.

## Installation

To install this package :

```bash
dart pub add salat
```

Then, run `dart pub get` to fetch the package.

## Usage
Here's a simple example demonstrating how to calculate prayer times using the prayer_times package:

```dart
import 'package:salat/salat.dart';
import 'package:timezone/standalone.dart' as tz;

void main(List<String> args) async {
  await tz.initializeTimeZone();

  String timezone = "Asia/Riyadh"; // Time zone of Mecca (Riyadh)
  final pt = prayerTimes(
      method: CalculationMethod.MAKKAH); // Create prayer times object

  double longitude = 39.857910; // Longitude of the location
  double latitude = 21.389082; // Latitude of the location

  // Calculate prayer times for the given location
  final prayertimes = pt.calcTime(
    date: DateTime.now(),
    timezone: timezone,
    longitude: longitude,
    latitude: latitude,
  );
  printPrayerTimes(prayertimes);
}
```
### output
| Name     | Time                       |
|----------|----------------------------|
| fajr     | July 14, 2023 04:20:49 +3  |
| sunrise  | July 14, 2023 05:47:00 +3  |
| dhuhr    | July 14, 2023 12:26:32 +3  |
| asr      | July 14, 2023 15:40:29 +3  |
| maghrib  | July 14, 2023 19:05:54 +3  |
| isha     | July 14, 2023 20:35:54 +3  |
| midnight | July 14, 2023 12:26:27 +3  |

Please refer to the documentation for more details on how to use this package.

## Contributing
Contributions are welcome! If you encounter any issues, have suggestions, or would like to contribute to the package, please feel free to open an issue or submit a pull request on the GitHub repository.


CHANGELOG.md
## 1.0.1
- readme fix.
## 1.0.2
- add next prayer feature
## 1.0.4
- assert fix

lib/src/calculations.dart
import 'dart:math' as math;

extension DurationExtensions on Duration {
  double toSeconds() {
    return inMicroseconds / Duration.microsecondsPerSecond;
  }
}

double radians(double degrees) {
  return degrees * math.pi / 180.0;
}

bool isClose(double a, double b) {
  double tolerance = 1e-6; // Example tolerance
  return (a - b).abs() < tolerance;
}

List eotDecl(DateTime time) {
  DateTime epoch = DateTime.utc(2000, 1, 1, 12);
  double days_since_epoch =
      (time.difference(epoch) + Duration(days: 1)).toSeconds() / 60 / 60 / 24;
  double y100 = days_since_epoch / 36525;
  double e = 1.6709e-2 - 4.193e-5 * y100 - 1.26e-7 * math.pow(y100, 2);
  double lam_p =
      radians(282.93807 + 1.7195 * y100 + 3.025e-4 * math.pow(y100, 2));
  double epsilon = radians(23.4393 -
      0.013 * y100 -
      2e-7 * math.pow(y100, 2) +
      5e-7 * math.pow(y100, 3));
  double MD = 6.24004077;
  double TY = 365.2596358;
  double D = days_since_epoch % TY;
  double M = MD + 2 * math.pi * D / TY;
  M = M % (2 * math.pi);
  double E = keplerSolve(M: M, e: e);
  double nu = math.acos((math.cos(E) - e) / (1 - e * math.cos(E)));
  if (E > math.pi) {
    nu = 2 * math.pi - nu;
  }
  double lam = nu + lam_p;
  lam = lam % (2 * math.pi);
  double alpha;
  if (isClose(math.cos(lam), 0)) {
    alpha = lam;
  } else {
    alpha = math.atan(math.cos(epsilon) * math.tan(lam));
    if (lam < math.pi / 2) {
      assert((0 <= alpha) && (alpha < math.pi / 2));
    } else if (lam < math.pi * 3 / 2) {
      alpha += math.pi;
      assert((math.pi / 2 <= alpha) && (alpha < math.pi * 3 /2));
    } else {
      alpha += 2 * math.pi;
      assert((math.pi * 3 / 2 <= alpha) && (alpha < 2 * math.pi));
    }
  }
  double eot_rad = M + lam_p - alpha;
  if (eot_rad > math.pi) {
    eot_rad -= 2 * math.pi;
  }
  double eot_min = eot_rad / (2 * math.pi) * 60 * 24;
  Duration eot = Duration(seconds: (eot_min * 60).round());
  double decl = math.asin(math.sin(epsilon) * math.sin(lam));
  return [eot, decl];
}

double keplerSolve({required double M, required double e}) {
  if (!(e > 0 && e < 1)) {
    throw Exception(
        "Eccentricity of elliptical orbit required in range (0, 1)");
  }
  double E = M;
  while (!isClose(M, E - e * math.sin(E))) {
    E = E - (E - e * math.sin(E) - M) / (1 - e * math.cos(E));
  }
  return E;
}

double calcAltitude(
    {required int shadowFactor,
    required double declination,
    required double latitude}) {
  double phi = radians(latitude);
  double delta = declination;
  double alt = math.atan(1 / (shadowFactor + math.tan(phi - delta)));
  return alt;
}

Duration timedeltaAtAltitude(
    {required double altitude,
    required double declination,
    required double latitude}) {
  double alpha = altitude;
  double phi = radians(latitude);
  double delta = declination;
  double numerator = -math.sin(alpha) - math.sin(phi) * math.sin(delta);
  double denominator = math.cos(phi) * math.cos(delta);
  double cosHourRad = numerator / denominator;
  if (cosHourRad < -1 || cosHourRad > 1) {
    throw Exception("Sun does not reach altitude");
  }
  double hourRad = math.acos(cosHourRad);
  double hours = hourRad / (2 * math.pi) * 24;
  Duration T = Duration(seconds: (hours * 60 * 60).round());

  assert((Duration.zero <= T) && (T <= Duration(hours: 12)));
  return T;
}

DateTime linearInterpolation(
    {required Duration Function(DateTime) diffFunction,
    required DateTime guess1,
    required DateTime guess2}) {
  if (isClose((guess1.difference(guess2).inSeconds.toDouble()), 0)) {
    throw Exception("guess1 and guess2 need to be different");
  }
  if (guess2.isBefore(guess1)) {
    DateTime temp = guess1;
    guess1 = guess2;
    guess2 = temp;
  }
  Duration diff1 = diffFunction(guess1);
  Duration diff2 = diffFunction(guess2);
  while (!isClose((guess1.difference(guess2).inSeconds.toDouble()), 0)) {
    Duration diff = guess2.difference(guess1);
    double ratio = diff.inMilliseconds / (diff2 - diff1).inMilliseconds;
    DateTime guess3 = guess1.subtract(diff1 * ratio);
    Duration diff3 = diffFunction(guess3);

    guess1 = guess2;
    diff1 = diff2;
    guess2 = guess3;
    diff2 = diff3;
  }
  return guess1;
}

DateTime timeZenith({required DateTime localNoon, required double longitude}) {
  DateTime utcNoonApprox =
      localNoon.add(Duration(seconds: (longitude / 15 * 3600).round()));
  utcNoonApprox = utcNoonApprox.toUtc();
  DateTime utcna = utcNoonApprox;
  DateTime utcNoon = DateTime.utc(utcna.year, utcna.month, utcna.day, 12);
  Duration calcDifference(DateTime guess) {
    Duration eot = eotDecl(guess)[0];

    DateTime actual = utcNoon
        .subtract(Duration(seconds: (longitude / 15 * 3600).round()))
        .subtract(eot);
    return actual.difference(guess);
  }

  DateTime guess =
      utcNoon.subtract(Duration(seconds: (longitude / 15 * 3600).round()));
  DateTime guess1 = guess.subtract(Duration(minutes: 20));
  DateTime guess2 = guess.add(Duration(minutes: 20));
  return linearInterpolation(
      diffFunction: calcDifference, guess1: guess1, guess2: guess2);
}

DateTime timeAltitude(
    {required DateTime localNoon,
    required double altitude,
    required double longitude,
    required double latitude,
    required bool rising}) {
  DateTime zenith = timeZenith(localNoon: localNoon, longitude: longitude);

  Duration calcDifference(DateTime guess) {
    double declination = eotDecl(guess)[1];
    Duration T = timedeltaAtAltitude(
        altitude: altitude, declination: declination, latitude: latitude);
    DateTime actual;
    if (rising) {
      actual = zenith.subtract(T);
    } else {
      actual = zenith.add(T);
    }
    return actual.difference(guess);
  }

  DateTime guess1;
  DateTime guess2;
  if (rising) {
    guess1 = zenith.subtract(Duration(hours: 12));
    guess2 = zenith;
  } else {
    guess1 = zenith;
    guess2 = zenith.add(Duration(hours: 12));
  }

  return linearInterpolation(
      diffFunction: calcDifference, guess1: guess1, guess2: guess2);
}

DateTime timeShadowFactor(
    {required DateTime localNoon,
    required int shadowFactor,
    required double longitude,
    required double latitude,
    required bool rising}) {
  DateTime zenith = timeZenith(localNoon: localNoon, longitude: longitude);

  Duration calcDifference(DateTime guess) {
    double declination = eotDecl(guess)[1];
    double altitude = calcAltitude(
        shadowFactor: shadowFactor,
        declination: declination,
        latitude: latitude);
    Duration T = timedeltaAtAltitude(
        altitude: -altitude, declination: declination, latitude: latitude);
    DateTime actual;
    if (rising) {
      actual = zenith.subtract(T);
    } else {
      actual = zenith.add(T);
    }
    return actual.difference(guess);
  }

  DateTime guess1;
  DateTime guess2;
  if (rising) {
    guess1 = zenith.subtract(Duration(hours: 12));
    guess2 = zenith;
  } else {
    guess1 = zenith;
    guess2 = zenith.add(Duration(hours: 12));
  }
  return linearInterpolation(
      diffFunction: calcDifference, guess1: guess1, guess2: guess2);
}

void main(List<String> args) {
  eotDecl(DateTime.now());
}


lib/src/services/prayertime_service.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first

// import 'package:geolocator/geolocator.dart';
import 'package:salat/salat.dart';

class Prayers {
  final DateTime _date;
  final String _timezone;
  final double _longitude;
  final double _latitude;
  final CalculationMethod _method;
  late DateTime fajr;
  late DateTime dhuhr;
  late DateTime asr;
  late DateTime maghrib;
  late DateTime isha;
  Prayers({
    required DateTime date,
    required String timezone,
    required double longitude,
    required double latitude,
    required CalculationMethod method,
  })  : _latitude = latitude,
        _longitude = longitude,
        _timezone = timezone,
        _method = method,
        _date = date {
    final pt = prayerTimes(method: method);
    final times = pt.calcTime(
      date: _date,
      timezone: _timezone,
      longitude: _longitude,
      latitude: _latitude,
    );
    fajr = times['fajr']!;
    dhuhr = times['dhuhr']!;
    asr = times['asr']!;
    maghrib = times['maghrib']!;
    isha = times['isha']!;
  }

  // determines the next prayer
  Future<Map<String, DateTime>> nextPrayer() async {
    final pt = prayerTimes(method: _method);
    final times = pt.calcTime(
      date: _date,
      timezone: _timezone,
      longitude: _longitude,
      latitude: _latitude,
    );
    final prayersList = times.entries.indexed;
    Map<String, DateTime> next = {};
    DateTime now = _date;
    for (int i = 0; i < prayersList.length; i++) {
      final element = prayersList.elementAt(i);
      final index = element.$1;

      // next prayer : fajr (now < fajr)
      if (now.isBefore(prayersList.elementAt(0).$2.value)) {
        final entry = prayersList.elementAt(0).$2;
        next = {entry.key: entry.value};
        break;
      }
      // next prayer : fajr (isha < now)
      else if (now
          .isAfter(prayersList.elementAt(prayersList.length - 1).$2.value)) {
        final entry = prayersList.elementAt(0).$2;
        next = {entry.key: entry.value};

        break;
      }
      // framing
      else if (now.isAfter(element.$2.value) &&
          now.isBefore(prayersList.elementAt(index + 1).$2.value)) {
        final entry = prayersList.elementAt(index + 1).$2;
        next = {entry.key: entry.value};
        break;
      }
    }
    return next;
  }

  @override
  String toString() {
    return 'Prayers(Fajr: $fajr, Dhuhr: $dhuhr, Asr: $asr, Maghrib: $maghrib, Isha: $isha)';
  }

  // describing itself :)
  Map<String, DateTime> timesToMap() {
    return <String, DateTime>{
      'fajr': fajr,
      'dhuhr': dhuhr,
      'asr': asr,
      'maghrib': maghrib,
      'isha': isha,
    };
  }
}

lib/src/methods.dart
// ignore_for_file: constant_identifier_names

import 'calculations.dart';
import 'package:timezone/timezone.dart';
import 'package:hijri/hijri_calendar.dart';

enum CalculationMethod {
  ISNA,
  MWL,
  EGYPT,
  TEHRAN,
  JAFARI,
  MAKKAH,
  KARACHI,
}

enum AsrMethod {
  STANDARD,
  HANAFI,
}

class GeneralMethod {
  double fajrAltitude;
  double ishaAltitude;
  double sunsetAltitude;
  AsrMethod asrMethod;
  late int shadowFactor;

  GeneralMethod(
      {required double fajrAltitudeDeg,
      required double ishaAltitudeDeg,
      this.asrMethod = AsrMethod.STANDARD})
      : shadowFactor = asrMethod == AsrMethod.STANDARD ? 1 : 2,
        fajrAltitude = radians(fajrAltitudeDeg),
        sunsetAltitude = radians(0.833),
        ishaAltitude = radians(ishaAltitudeDeg);
  Map<String, DateTime> calcTime(
      {required DateTime date,
      required String timezone,
      required double longitude,
      required double latitude}) {
    DateTime localNoon =
        TZDateTime(getLocation(timezone), date.year, date.month, date.day, 12);
    DateTime fajr = timeAltitude(
        localNoon: localNoon,
        altitude: fajrAltitude,
        longitude: longitude,
        latitude: latitude,
        rising: true);
    DateTime sunrise = timeAltitude(
        localNoon: localNoon,
        altitude: sunsetAltitude,
        longitude: longitude,
        latitude: latitude,
        rising: true);
    DateTime dhuhr = timeZenith(
      localNoon: localNoon,
      longitude: longitude,
    );
    DateTime asr = timeShadowFactor(
        localNoon: localNoon,
        shadowFactor: shadowFactor,
        longitude: longitude,
        latitude: latitude,
        rising: false);
    DateTime maghrib = timeAltitude(
        localNoon: localNoon,
        altitude: sunsetAltitude,
        longitude: longitude,
        latitude: latitude,
        rising: false);
    DateTime isha = timeAltitude(
        localNoon: localNoon,
        altitude: ishaAltitude,
        longitude: longitude,
        latitude: latitude,
        rising: false);
    DateTime sunset = maghrib;
    DateTime nextLocalNoon = localNoon.add(Duration(days: 1));
    DateTime nextSunrise = timeAltitude(
      localNoon: nextLocalNoon,
      altitude: sunsetAltitude,
      longitude: longitude,
      latitude: latitude,
      rising: true,
    );
    DateTime midnight = sunset.add((Duration(
        milliseconds:
            (nextSunrise.difference(sunset).inMilliseconds / 2).round())));

    Map<String, DateTime> times = {
      "fajr": fajr,
      "sunrise": sunrise,
      "dhuhr": dhuhr,
      "asr": asr,
      "maghrib": maghrib,
      "isha": isha,
      "midnight": midnight,
    };
    times.forEach((key, value) {
      times[key] = TZDateTime.from(value, getLocation(timezone));
    });
    return times;
  }
}

class TehranMethod extends GeneralMethod {
  final AsrMethod asrMethod = AsrMethod.STANDARD;
  TehranMethod()
      : super(
            fajrAltitudeDeg: 17.7,
            ishaAltitudeDeg: 14,
            asrMethod: AsrMethod.STANDARD);
  @override
  Map<String, DateTime> calcTime(
      {required DateTime date,
      required String timezone,
      required double longitude,
      required double latitude}) {
    DateTime localNoon =
        TZDateTime(getLocation(timezone), date.year, date.month, date.day, 12);
    double magribAltitude = radians(4.5);
    Map<String, DateTime> times = super.calcTime(
        date: date,
        timezone: timezone,
        longitude: longitude,
        latitude: latitude);
    DateTime sunset = times["maghrib"]!;
    DateTime maghrib = timeAltitude(
        localNoon: localNoon,
        altitude: magribAltitude,
        longitude: longitude,
        latitude: latitude,
        rising: false);
    times["maghrib"] = maghrib;
    DateTime nextLocalNoon = localNoon.add(Duration(days: 1));
    DateTime nextFajr = timeAltitude(
        localNoon: nextLocalNoon,
        altitude: fajrAltitude,
        longitude: longitude,
        latitude: latitude,
        rising: true);
    DateTime midnight = sunset.add((Duration(
        milliseconds:
            (nextFajr.difference(sunset).inMilliseconds / 2).round())));
    times["midnight"] = midnight;
    return times;
  }
}

class JafariMethod extends GeneralMethod {
  final AsrMethod asrMethod = AsrMethod.STANDARD;
  JafariMethod()
      : super(
            fajrAltitudeDeg: 16,
            ishaAltitudeDeg: 14,
            asrMethod: AsrMethod.STANDARD);
  Map<String, DateTime> calcTime(
      {required DateTime date,
      required String timezone,
      required double longitude,
      required double latitude}) {
    DateTime localNoon =
        TZDateTime(getLocation(timezone), date.year, date.month, date.day, 12);
    double magribAltitude = radians(4);
    Map<String, DateTime> times = super.calcTime(
        date: date,
        timezone: timezone,
        longitude: longitude,
        latitude: latitude);
    DateTime sunset = times["maghrib"]!;
    DateTime maghrib = timeAltitude(
        localNoon: localNoon,
        altitude: magribAltitude,
        longitude: longitude,
        latitude: latitude,
        rising: false);
    times["maghrib"] = maghrib;
    DateTime nextLocalNoon = localNoon.add(Duration(days: 1));
    DateTime nextFajr = timeAltitude(
        localNoon: nextLocalNoon,
        altitude: fajrAltitude,
        longitude: longitude,
        latitude: latitude,
        rising: true);
    DateTime midnight = sunset.add((Duration(
        milliseconds:
            (nextFajr.difference(sunset).inMilliseconds / 2).round())));
    times["midnight"] = midnight;
    return times;
  }
}

class MakkahMethod extends GeneralMethod {
  final AsrMethod asrMethod = AsrMethod.STANDARD;
  MakkahMethod()
      : super(
            fajrAltitudeDeg: 18.5,
            ishaAltitudeDeg: 18.5,
            asrMethod: AsrMethod.STANDARD);
  @override
  Map<String, DateTime> calcTime(
      {required DateTime date,
      required String timezone,
      required double longitude,
      required double latitude}) {
    Map<String, DateTime> times = super.calcTime(
        date: date,
        timezone: timezone,
        longitude: longitude,
        latitude: latitude);
    HijriCalendar hijriDate =
        HijriCalendar.fromDate(DateTime(date.year, date.month, date.day));
    if (hijriDate.hMonth == 9) {
      times["isha"] = times["maghrib"]!.add(Duration(minutes: 120));
    } else {
      times["isha"] = times["maghrib"]!.add(Duration(minutes: 90));
    }
    return times;
  }
}

GeneralMethod prayerTimes(
    {CalculationMethod method = CalculationMethod.MWL,
    AsrMethod asr = AsrMethod.STANDARD}) {
  if (method == CalculationMethod.ISNA) {
    return GeneralMethod(fajrAltitudeDeg: 15, ishaAltitudeDeg: 15);
  } else if (method == CalculationMethod.MWL) {
    return GeneralMethod(fajrAltitudeDeg: 18, ishaAltitudeDeg: 17);
  } else if (method == CalculationMethod.EGYPT) {
    return GeneralMethod(fajrAltitudeDeg: 19.5, ishaAltitudeDeg: 17.5);
  } else if (method == CalculationMethod.KARACHI) {
    return GeneralMethod(fajrAltitudeDeg: 18, ishaAltitudeDeg: 18);
  } else if (method == CalculationMethod.TEHRAN) {
    return TehranMethod();
  } else if (method == CalculationMethod.JAFARI) {
    return JafariMethod();
  } else if (method == CalculationMethod.MAKKAH) {
    return MakkahMethod();
  } else {
    throw Exception("Unknown CalculationMethod ${method.toString()}");
  }
}
